src/snapshot/src/persist.rs

```diff
@@ -0,0 +1,25 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//! Defines an abstract interface for saving/restoring a component from state.
+#![deny(missing_docs)]
+
+/// An abstract interface for saving/restoring a component from state.
+pub trait Persist
+where
+    Self: Sized,
+{
+    /// The type of the object representing the state of the component.
+    type State;
+    /// The type of the object holding the constructor arguments.
+    type ConstructorArgs;
+    /// The type of the error that can occur while constructing the object.
+    type Error;
+
+    /// Returns the current state of the component.
+    fn save(&self) -> Self::State;
+    /// Constructs a component from a specified state.
+    fn restore(
+        constructor_args: Self::ConstructorArgs,
+        state: &Self::State,
```

Comment:
Could we pass ownership of the State, or do we need to use the State after restoring the component?

Reply:
Usually that state will be part of a bigger structure, so we will pass a reference in the general case as we expect these to be larger.